File: M79.py
import copy

class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        found = False
        taken = set()
        for i in range(len(board)):
            for j in range(len(board[0])):
                if not found and board[i][j] == word[0]:
                    new_taken = copy.copy(taken)
                    found = self.helper(board, word, i, j, 1, new_taken)
        return found

    def helper(self, board, word, i, j, word_pos, taken):
        if word_pos >= len(word):
            return True
            
        taken.add((i, j))

        found = False
        if not found and (i+1, j) not in taken and i + 1 < len(board) and board[i+1][j] == word[word_pos]:
            new_taken = copy.copy(taken)
            found = self.helper(board, word, i+1, j, word_pos + 1, new_taken)
        if not found and (i, j+1) not in taken and j + 1 < len(board[0]) and board[i][j+1] == word[word_pos]:
            new_taken = copy.copy(taken)
            found = self.helper(board, word, i, j+1, word_pos + 1, new_taken)
        if not found and (i-1, j) not in taken and i - 1 >= 0 and board[i-1][j] == word[word_pos]:
            new_taken = copy.copy(taken)
            found = self.helper(board, word, i-1, j, word_pos + 1, new_taken)
        if not found and (i, j-1) not in taken and j - 1 >= 0 and board[i][j-1] == word[word_pos]:
            new_taken = copy.copy(taken)
            found = self.helper(board, word, i, j-1, word_pos + 1, new_taken)
        return found

File: M79.cc
#include <iostream>
#include <map>
#include <utility>

using namespace std;

class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        vector<pair<int, int>> taken;
        bool found = false;
        for (int y = 0; y < size(board); y++) {
            for (int x = 0; x < size(board[0]); x++) {
                if (!found && board[y][x] == word[0]){
                    found = helper(board, y, x, word, 1, taken);
                }
            }
        }
        return found;
    }

    bool helper(vector<vector<char>>& board, int curr_y, int curr_x, string& word, int word_pos, vector<pair<int, int>> taken) {

        if (word_pos == size(word)) {
            return true;
        }
        taken.push_back(make_pair(curr_x, curr_y));
        bool found = false;
        if (!found && curr_y+1 < size(board) && board[curr_y+1][curr_x] == word[word_pos] && find(taken.begin(), taken.end(), make_pair(curr_x, curr_y+1)) == taken.end()) {
            found = helper(board, curr_y+1, curr_x, word, word_pos + 1, taken);
        }
        if (!found && curr_y-1 >= 0 && board[curr_y-1][curr_x] == word[word_pos] && find(taken.begin(), taken.end(), make_pair(curr_x, curr_y-1)) == taken.end()) {
            found = helper(board, curr_y-1, curr_x, word, word_pos + 1, taken);
        }
        if (!found && curr_x+1 < size(board[0]) && board[curr_y][curr_x+1] == word[word_pos] && find(taken.begin(), taken.end(), make_pair(curr_x+1, curr_y)) == taken.end()) {
            found = helper(board, curr_y, curr_x+1, word, word_pos + 1, taken);
        }
        if (!found && curr_x-1 >= 0 && board[curr_y][curr_x-1] == word[word_pos] && find(taken.begin(), taken.end(), make_pair(curr_x-1, curr_y)) == taken.end()) {
            found = helper(board, curr_y, curr_x-1, word, word_pos + 1, taken);
        }
        return found;
    }
};